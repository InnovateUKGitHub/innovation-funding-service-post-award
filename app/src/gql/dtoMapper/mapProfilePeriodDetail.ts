// on Acc_Profile__c WHERE Acc_ProjectID__c=ProjectID AND RecordType.Name="Total Project Period"
type ProfilePeriodDetailsNode = Readonly<
  Partial<{
    Acc_PeriodLatestForecastCost__c: GQL.Value<number>;
    Acc_ProjectParticipant__c?: GQL.Value<string>;
    Acc_ProjectPeriodNumber__c: GQL.Value<number>;
  }>
> | null;

type ProfilePeriodDetailsDtoMapping = { partnerId: PartnerId; forecastCost: number; periodId: number };

const mapper: GQL.DtoMapper<ProfilePeriodDetailsDtoMapping, ProfilePeriodDetailsNode> = {
  partnerId(node) {
    return (node?.Acc_ProjectParticipant__c?.value ?? "") as PartnerId;
  },
  forecastCost(node) {
    return node?.Acc_PeriodLatestForecastCost__c?.value ?? 0;
  },
  periodId(node) {
    return node?.Acc_ProjectPeriodNumber__c?.value ?? 0;
  },
};

/**
 * Maps a specified ProfilePeriodDetails Node from a GQL query to a slice of
 * the ProfilePeriodDetails to ensure consistency and compatibility in the application
 */
export function mapToProfilePeriodDetailsDto<
  T extends ProfilePeriodDetailsNode,
  PickList extends keyof ProfilePeriodDetailsDtoMapping,
>(node: T, pickList: PickList[]): Pick<ProfilePeriodDetailsDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](node);
    return dto;
  }, {} as Pick<ProfilePeriodDetailsDtoMapping, PickList>);
}

/**
 * Maps ProfilePeriodDetails Edge to array of Loan DTOs.
 */
export function mapToProfilePeriodDetailsDtoArray<
  T extends ReadonlyArray<{ node: ProfilePeriodDetailsNode } | null> | null,
  PickList extends keyof ProfilePeriodDetailsDtoMapping,
>(edges: T, pickList: PickList[]): Pick<ProfilePeriodDetailsDtoMapping, PickList>[] {
  return (
    edges?.map(node => {
      return mapToProfilePeriodDetailsDto(node?.node ?? null, pickList);
    }) ?? []
  );
}
