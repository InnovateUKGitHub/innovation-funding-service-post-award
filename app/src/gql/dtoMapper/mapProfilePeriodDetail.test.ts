import { mapToProfilePeriodDetailsDtoArray } from "./mapProfilePeriodDetail";

const edges = [
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 10,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 20,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 30,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 40,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 50,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 60,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 70,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 80,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 90,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
  {
    node: {
      Acc_PeriodLatestForecastCost__c: {
        value: 100,
      },
      Acc_ProjectParticipant__c: {
        value: "a0D2600000zXl70EAC",
      },
    },
  },
];

describe("mapProfilePeriodDetailDtoArray", () => {
  it("should map the gql node to the dto array", () => {
    expect(mapToProfilePeriodDetailsDtoArray(edges, ["forecastCost", "partnerId"])).toMatchSnapshot();
  });
});
