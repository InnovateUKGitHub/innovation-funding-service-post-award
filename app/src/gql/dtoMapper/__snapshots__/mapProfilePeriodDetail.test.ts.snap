// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapProfilePeriodDetailDtoArray should map the gql node to the dto array 1`] = `
[
  {
    "forecastCost": 10,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 20,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 30,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 40,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 50,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 60,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 70,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 80,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 90,
    "partnerId": "a0D2600000zXl70EAC",
  },
  {
    "forecastCost": 100,
    "partnerId": "a0D2600000zXl70EAC",
  },
]
`;
